keps/0010-dynamic-plugin-configuration.md

```diff
@@ -0,0 +1,221 @@
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories (Optional)](#user-stories-optional)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+    - [Story 3](#story-3)
+    - [Story 4](#story-4)
+    - [Story 5](#story-5)
+  - [Notes/Constraints/Caveats (Optional)](#notesconstraintscaveats-optional)
+    - [Notes](#notes)
+    - [Caveats](#caveats)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+Currently the configuration for the `KongPlugin` and `KongClusterPlugin` resources
+is generally static, meaning that it is not possible to derive configuration values from
+common configuration resources defined on same kubernetes cluster. This leads to a lot
+duplication of hard coded values in numerous places making the deployment process
+difficult to manage. While it is possible to load configuration from a secret using
+the `configFrom` parameter, the entire configuration must be contained with in the
+secret which is effectively the same problem.
+
+It would be advantageous, and more in line with common Kubernetes behavior to be able
+to derive individual values of plugin configuration from data already stored on the same
+Kubernetes cluster. This proposal aims to define a change to the `KongPlugin` and `KongClusterPlugin`
+Custom Resource Definitions to facilitate this behavior.
+
+## Motivation
+
+Allow `KongPlugin` and `KongClusterPlugin` resources to read configuration values from Kubernetes,
+in addition to literal values.
+
+Continue to allow `KongPlugin` and `KongClusterPlugin` resource to accept literal input values.
+
+Reduce the amount of duplication of configuration value in the simple case. That is,
+when using standard kubernetes objects and with out assumed templating systems.
+
+### Goals
+
+* Provide ability to read individual configuration values from `ConfigMap` resources
+* Provide ability to read individual configuration values from `Secret` resources
+* Provide ability to read individual configuration values from arbitrary or generic resource types (other custom resource types)
+
+### Non-Goals
+
+* Introduction of a new Custom Resource or different Plugin Resource
+
+## Proposal
+
+In an effort to stay in line with current Kubernetes resource patterns it would be desirable
+to mimic the way pods define environment variables using an array of objects where
+the objects contain a `name` and one of `value` or `valueFrom`. `valueFrom`
+would further supports `configMapKeyRef`, `secretKeyRef` or `genericKeyRef`
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a operator I want certain plugin configuration values to be dynamic in addition
+to statically definable
+
+#### Story 2
+
+As a operator I want to be able to derive a literal value using value interpolation referencing
+other values
+
+#### Story 3
+
+As an operator I want values stored in `ConfigMap` or `Secret` resources
+to be the source of truth for various resources deployed to my Kubernetes cluster
```

Comment:
I'm not sure I'm understanding this user story at a glance, is this paraphrase a good interpretation?

```suggestion
As an operator I want to be able to store common plugin options and configurations as
`ConfigMaps` or `Secrets` such that I can write common configurations once and not
have to re-write them across multiple plugins.
```

Reply:
Yes. I suppose I was trying to indicate that we want to be able to leverage configuration that already exists on the cluster and that isn't put there specifically for plugins. re-use across resources on a cluster.

But, yes this captures it in the context of what the controller cares about.